torch/multiprocessing/reductions.py

```diff
@@ -125,24 +128,63 @@ def reduce_tensor(tensor):
     # *entire* cudaMalloc allocation, i.e., the 0xA000 region, not just
     # the storage 0xA100 (because that is what CUDA supports).  So, on the
     # other end, there simply isn't any way to say, "Wait, you gave me
-    # a bigger region (0xA000) than the one I wanted (0xA100)"; we have
-    # to just make a storage for the entire caching allocator block.
+    # a bigger region (0xA000) than the one I wanted (0xA100)".
+    #
+    # Note that this cudaMalloc allocation might not be a single type of storage.
+    # It's possible that it contains storages of mixed types: float, bytes...
+    # Thus we cannot simply make a single storage out of the allocation.
+    #
+    # cudaIpcMemHandle is an identifier to access the sender memory on the
+    # receiver side. However, cudaIpcMemHandles from each device in a given process may
+    # only be opened by one context per device per other process.
+    # This requires us to cache the device pointer for each cudaIpcMemHandle on C++
+    # side to reconstruct types of storages, while keep the old ones alives.
+    # See [https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__DEVICE.html]
     #
-    # This is fine, because all we need to do is just adjust the offset
-    # on the tensor itself: instead of:
+    # This is fine, because all we need to do is to save our position in the allocaiton,
+    # and reconstruct storage and tensor from it.
+    # 0xA000 ->  -------CUDA Allocation------
+    #           |                            |
+    #           |                            |
+    #           |                            |
+    #           |                            |
+    # 0xA100 ->  --------storage1 begin------
+    #           |                            |
+    # 0xA120 ->  --------tensor1 begin ------
+    #           |                            |
+    #           |                            |
+    #           |                            |
+    #           |                            |
+    #           |                            |
+    # 0xA160 ->  --------tensor1 end---------
+    #           |                            |
+    #           |                            |
+    #           |                            |
+    # 0xA200 ->  --------storage1 end--------
+    #           |                            |
+    # 0xE000 ->  --------CUDA allocation-----
     #
-    #   Tensor(size=0x100, offset=0x020, storage=Storage(data=0xA100, size=0x0100))
+    # To send a tensor
+    #   Tensor(size=0x040, offset=0x020, storage=Storage(data=0xA100, size=0x0100))
```

Comment:
I don't think this accurately describes what we do anymore. We don't send a "Storage", we send the CUDA allocation memory handle, and the offset of the storage into that allocation.